Mathlib/MeasureTheory/Measure/SeparableMeasure.lean

```diff
@@ -0,0 +1,494 @@
+The term "separable" is justified by the fact that the definition translates to the usual notion
+of separability in the metric space made by constant indicators equipped with the `L¹` norm.
+
+## Main definitions
+
+* `MeasureTheory.Measure.μ.MeasureDense 𝒜`: `𝒜` is a measure-dense family if it only contains
+  measurable sets and if the following condition is satisfied: if `s` is measurable with finite
+  measure, then for any `ε > 0` there exists `t ∈ 𝒜` such that `μ (s ∆ t) < ε `.
+* `MeasureTheory.IsSeparable`: A measure is separable if there exists a countable and
+  measure-dense family.
+
+## Main statements
+
+* `MeasureTheory.instSecondCountableLp`: If `μ` is separable, `E` is second-countable and
+  `1 ≤ p < ∞` then `Lp E p μ` is second-countable. This is in particular true if `X` is countably
+  generated and `μ` is s-finite.
+
+## Implementation notes
+
+Through the whole file we consider a measurable space `X` equipped with a measure `μ`, and also
+a normed commutative group `E`. We also consider an extended non-negative real `p` such that
+`1 ≤ p < ∞`. This is registered as instances via `one_le_p : Fact (1 ≤ p)` and
+`p_ne_top : Fact (p ≠ ∞)`, so the properties are accessible via `one_le_p.elim` and `p_ne_top.elim`.
+
+Through the whole file, when we write that an extended non-negative real is finite, it is always
+written `≠ ∞` rather than `< ∞`. See `Ne.lt_top` and `ne_of_lt` to switch from one to the other.
+
+## References
+
+* [D. L. Cohn, *Measure Theory*][cohn2013measure]
+
+## Tags
+
+separable measure, measure-dense, Lp space, second-countable
+-/
+
+open MeasurableSpace Set ENNReal TopologicalSpace BigOperators symmDiff Filter Real
+
+namespace MeasureTheory
+
+variable {X E : Type*} [m : MeasurableSpace X] [NormedAddCommGroup E] {μ : Measure X}
+variable {p : ℝ≥0∞} [one_le_p : Fact (1 ≤ p)] [p_ne_top : Fact (p ≠ ∞)] {𝒜 : Set (Set X)}
+
+section MeasureDense
+
+/-! ### Definition of a measure-dense family, basic properties and sufficient conditions -/
+
+/-- A family `𝒜` of sets of a measure space is said to be measure-dense if it contains only
+measurable sets and can approximate any measurable set with finite measure, in the sense that
+for any measurable set `s` with finite measure the symmetric difference `s ∆ t` can be made
+arbitrarily small when `t ∈ 𝒜`. We show below that such a family can be chosen to contain only
+sets with finite measures.
+
+The term "measure-dense" is justified by the fact that the approximating condition translates
+to the usual notion of density in the metric space made by constant indicators of measurable sets
+equipped with the `L¹` norm. -/
+structure Measure.MeasureDense (μ : Measure X) (𝒜 : Set (Set X)) : Prop :=
+  /-- Each set has to be measurable. -/
+  measurable : ∀ s ∈ 𝒜, MeasurableSet s
+  /-- Any measurable set can be approximated by sets in the family. -/
+  approx : ∀ s, MeasurableSet s → μ s ≠ ∞ → ∀ (ε : ℝ),
```

Comment:
Is there a reason why you are using a real number instead of an ennreal here? Since everything happens in ennreal, it would seem more natural.

Reply:
I did it this way because it felt more natural because on the paper $\epsilon$ is rarely infinite. If I make it an ennreal it will maybe simplify calculations, but also add a case because I can't deal with it the same way if it is infinite. I'll try so see what I get.

Reply:
I tried to use an ennreal or an nnreal, but I don't think it's a good idea. A real epsilon is useful because I use it to do real analysis. If it is ennreal, I don't have a metric space. If it is nnreal, then it's the same as using a real, because I can't lift the measure, otherwise it loses its properties (at least I don't know how to do it).